rfcs/text/0005_route_handler.md

```diff
@@ -0,0 +1,184 @@
+- Start Date: 2019-06-29
+- RFC PR: (leave this empty)
+- Kibana Issue: https://github.com/elastic/kibana/issues/33779
+
+# Summary
+
+Http Service in New platform should provide the ability to execute some logic in response to an incoming request and send the result of this operation back.
+
+# Basic example
+Declaring a route handler for `/url` endpoint:
+```typescript
+router.get(
+  { path: '/url', ...[otherRouteParameters] },
+  (context: Context, request: KibanaRequest, t: KibanaResponseToolkit) => {
+   // logic to handle request ...
+   return t.ok(result);
+);
+
+```
+
+# Motivation
+The new platform is built with library-agnostic philosophy and we cannot transfer the current solution for Network layer from Hapi. To avoid vendor lock-in in the future, we have to define route handler logic and request/response objects formats that can be implemented in any low-level library such as Express, Hapi, etc. It means that we are going to operate our own abstractions for such Http domain entities as Router, Route, Route Handler, Request, Response.
+
+# Detailed design
+The new platform doesn't support the Legacy platform `Route Handler` format nor exposes implementation details, such as [Hapi.ResponseToolkit](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/hapi/v17/index.d.ts#L984).
+Rather `Route Handler` in New platform has the next signature:
+```typescript
+type RequestHandler = (
+  context: Context,
+  request: KibanaRequest,
+  t: KibanaResponseToolkit
+) => KibanaResponse | Promise<KibanaResponse>;
+```
+and accepts next Kibana specific parameters as arguments:
+- context: [Context](https://github.com/elastic/kibana/blob/master/rfcs/text/0003_handler_interface.md). A handler context contains core service and plugin functionality already scoped to the incoming request.
+- request: [KibanaRequest](https://github.com/elastic/kibana/blob/master/src/core/server/http/router/request.ts). An immutable representation of the incoming request details, such as body, parameters, query, url and route information. Note: you **must** to specify route schema during route declaration to have access to `body, parameters, query` in the request object. You  cannot extend KibanaRequest with arbitrary data nor remove any properties from it.
+```typescript
+interface KibanaRequest {
+  url: url.Url;
+  headers: Record<string, string | string [] | undefined>;
```

Comment:
What is `undefined` user for? Will it be used to remove a header?

Reply:
What does header with an array as value `string []` mean?

Reply:
that's how node.js parses headers. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/v10/http.d.ts#L8

> What is undefined user for? Will it be used to remove a header?

There is no guarantee that all headers are present in an incoming request.

> What does header with an array as value string [] mean?

the same reasons as here https://github.com/elastic/kibana/pull/39767#discussion_r298959784